inference/__init__.py

```diff
@@ -0,0 +1,492 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from enum import IntEnum
+
+import numpy as np
+
+import os
+
+import torch
+import spqr_cuda
+
+from torch import Tensor as T
+torch.set_grad_enabled(False)
+
+import inference
+
+
+
+# Utility functions
+
+def list_flatten(W):
+    if torch.is_tensor(W):
+        return W.flatten()
+    else:
+        return torch.cat(W).flatten()
+
+
+def updiv(x, y): return (x + y - 1) // y
+
+
+
+# Data structures
+class SPQRUncompressed:
+    m: int
+    n: int
+    bits: int
+    W: T
+    beta1: int
+    beta2: int
+    W_s: T
+    W_z: T
+    W_s_s: T
+    W_s_z: T
+    W_z_s: T
+    W_z_z: T
+    row_offsets: T
+    col_vals: T
+    dense_row_count: int
+    row_ids: T
+    buff0: T
+
+    @property
+    def nnz(self) -> int:
+        return self.col_vals.shape[0]
+
+    def __init__(self, m, n, bits, W, beta1, beta2, W_s, W_z, W_s_s, W_s_z, W_z_s, W_z_z, row_offsets, col_ids, values,
+                 in_perm=None, out_perm=None, sparse_compression_strategy: int = 0):
+        self.m = m
+        self.n = n
+        self.bits = bits
+        self.W = W
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.W_s = W_s
+        self.W_z = W_z
+        self.W_s_s = W_s_s
+        self.W_s_z = W_s_z
+        self.W_z_s = W_z_s
+        self.W_z_z = W_z_z
+        self.row_offsets = row_offsets
+        self.col_ids = col_ids
+        self.values = values
+        self.in_perm = in_perm
+        self.out_perm = out_perm
+        self.buff0 = allocate_compressed_buffers(m, n, beta1, beta2,  self.W.device)
+
+
+        if self.values.shape[0] != 0:
+            self.col_vals = values.view(torch.int16).to(torch.int64).bitwise_left_shift(16).bitwise_or(
+                col_ids.view(torch.int16).to(torch.int64)).to(torch.int32)
+        else:
+            self.col_vals = torch.zeros(0)
+
+        if sparse_compression_strategy == 0:
```

Comment:
It would better instead of 0 and one to give the name of  sparse_compression_strategy in string.

Reply:
Refactored into an enum.